libyul/optimiser/BlockHasher.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+template<size_t N>
+static constexpr uint64_t compileTimeLiteralHash(char const (&_literal)[N])
+{
+	return compileTimeLiteralHash(_literal, N);
+}
+}
+
+std::map<Block const*, uint64_t> BlockHasher::run(Block const& _block)
+{
+	std::map<Block const*, uint64_t> result;
+	BlockHasher blockHasher(result);
+	blockHasher(_block);
+	return result;
+}
+
+void BlockHasher::operator()(Literal const& _literal)
+{
+	hash64(compileTimeLiteralHash("Literal"));
+	hash64(_literal.value.hash());
+	hash64(_literal.type.hash());
+	hash8(static_cast<uint8_t>(_literal.kind));
+}
+
+void BlockHasher::operator()(Identifier const& _identifier)
+{
+	hash64(compileTimeLiteralHash("Identifier"));
+	auto it = m_variableReferences.find(_identifier.name);
+	if (it == m_variableReferences.end())
+	{
+		it = m_variableReferences.emplace(_identifier.name, VariableReference {
+			m_externalIdentifierCount++,
+			true
+		}).first;
+		m_externalReferences.emplace_back(_identifier.name);
+	}
+	if (it->second.isExternal)
+		hash64(compileTimeLiteralHash("external"));
+	else
+		hash64(compileTimeLiteralHash("internal"));
+	hash64(it->second.id);
+}
+
+void BlockHasher::operator()(FunctionalInstruction const& _instr)
+{
+	hash64(compileTimeLiteralHash("FunctionalInstruction"));
+	hash8(static_cast<std::underlying_type_t<eth::Instruction>>(_instr.instruction));
+	hash64(_instr.arguments.size());
+	ASTWalker::operator()(_instr);
+}
+
+void BlockHasher::operator()(FunctionCall const& _funCall)
+{
+	hash64(compileTimeLiteralHash("FunctionCall"));
+	hash64(_funCall.functionName.name.hash());
+	hash64(_funCall.arguments.size());
+	ASTWalker::operator()(_funCall);
+}
+
+void BlockHasher::operator()(ExpressionStatement const& _statement)
+{
+	hash64(compileTimeLiteralHash("ExpressionStatement"));
+	ASTWalker::operator()(_statement);
+}
+
+void BlockHasher::operator()(Assignment const& _assignment)
+{
+	hash64(compileTimeLiteralHash("Assignment"));
+	hash64(_assignment.variableNames.size());
+	for (auto const& name: _assignment.variableNames)
+		(*this)(name);
+	visit(*_assignment.value);
+}
+
+void BlockHasher::operator()(VariableDeclaration const& _varDecl)
+{
+	hash64(compileTimeLiteralHash("VariableDeclaration"));
+	hash64(_varDecl.variables.size());
+	for (auto const& var: _varDecl.variables)
+	{
+		yulAssert(!m_variableReferences.count(var.name), "");
+		m_variableReferences[var.name] = VariableReference{
+			m_internalIdentifierCount++,
+			false
+		};
+	}
+	ASTWalker::operator()(_varDecl);
+}
+
+void BlockHasher::operator()(If const& _if)
+{
+	hash64(compileTimeLiteralHash("If"));
+	ASTWalker::operator()(_if);
+}
+
+void BlockHasher::operator()(Switch const& _switch)
+{
+	hash64(compileTimeLiteralHash("Switch"));
+	hash64(_switch.cases.size());
+	// Instead of sorting we could consider to combine
+	// the case hashes using a commutative operation here.
+	std::set<Case const*, SwitchCaseSorter> cases;
```

Comment:
```suggestion
	std::set<Case const*, SwitchCaseSorter> cases(begin(_switch.cases), end(_switch.cases));
```

Reply:
Wanted to do that as well first, but ``_switch.cases`` is a ``std::vector<Case>`` and I don't want to copy the cases, but take their addresses and I'd say constructing transform iterators here is not really nice either...